File: java-samples/Tokens/tokentofriend/contracts/src/main/java/net/corda/samples/tokentofriend/contracts/CustomTokenContract.java
package net.corda.samples.tokentofriend.contracts;

import com.r3.corda.lib.tokens.contracts.EvolvableTokenContract;
import net.corda.samples.tokentofriend.states.CustomTokenState;
import net.corda.core.contracts.Contract;
import net.corda.core.transactions.LedgerTransaction;
import org.jetbrains.annotations.NotNull;
import static net.corda.core.contracts.ContractsDSL.requireThat;

public class CustomTokenContract extends EvolvableTokenContract implements Contract {

    public static final String CONTRACT_ID = "net.corda.samples.tokentofriend.contracts.CustomTokenContract";


    @Override
    public void additionalCreateChecks(@NotNull LedgerTransaction tx) {
        CustomTokenState newToken = (CustomTokenState) tx.getOutputStates().get(0);
        requireThat( require -> {
            require.using("Recipient Email cannot be empty",(!newToken.getReceipient().equals("")));
            require.using("Token Issuer Email and token recipient Email cannot be the same",
                    (!newToken.getReceipient().equals(newToken.getIssuer())));
            return null;
        });
    }

    @Override
    public void additionalUpdateChecks(@NotNull LedgerTransaction tx) {
        /*This additional check does not apply to this use case.
         *This sample does not allow token update */
    }
}

File: kotlin-samples/Tokens/tokentofriend/contracts/src/main/kotlin/net/corda/samples/tokentofriend/contracts/CustomTokenContract.kt
package net.corda.samples.tokentofriend.contracts

import com.r3.corda.lib.tokens.contracts.EvolvableTokenContract
import net.corda.samples.tokentofriend.states.CustomTokenState
import net.corda.core.contracts.Contract
import net.corda.core.transactions.LedgerTransaction

// ************
// * Contract *
// ************
class CustomTokenContract: EvolvableTokenContract(), Contract{

    companion object {
        @JvmStatic
        val CONTRACT_ID = "net.corda.samples.tokentofriend.contracts.CustomTokenContract"
    }

    override fun additionalCreateChecks(tx: LedgerTransaction) {
        val newToken = tx.outputStates.single() as CustomTokenState
        newToken.apply {
            require(recipient != "") {"Recipient Email cannot be empty"}
            require(recipient != issuer) {"Token Issuer Email and token recipient Email cannot be the same"}
        }
    }

    override fun additionalUpdateChecks(tx: LedgerTransaction) {
        /*This additional check does not apply to this use case.
         *This sample does not allow token update */
    }
}